the-keelson-protocol.md

```diff
@@ -0,0 +1,87 @@
+# The keelson protocol
+
+In short, keelson has opinions about:
+* The format of the key used when publishing data to zenoh
+* The format of the data published to zenoh
+* The format of the key used when declaring a queryable (i.e. RPC endpoint) in zenoh
+* The format of the requests and responses exchanged via a queryable (i.e. RPC endpoint) in zenoh
+
+**What is Zenoh?**
+
+In order to ease the introduction to keelson, make sure you are aquainted with zenoh. The following are some good resources:
+* [What is Zenoh?](https://zenoh.io/docs/overview/what-is-zenoh/)
+* [Zenoh in action](https://zenoh.io/docs/overview/zenoh-in-action/)
+* [The basic abstractions](https://zenoh.io/docs/manual/abstractions/)
+* [Zenoh: Unifying Communication, Storage and
+Computation from the Cloud to the Microcontroller](https://drive.google.com/file/d/1ETSLz2ouJ2o9OpVvEoXrbGcCvpF4TwJy/view?pli=1)
+
+## 1. Common key-space design
+
+In zenoh, both pub/sub and req/rep (queryables) messaging patterns all live in the same shared key "space". In keelson, the shared key-space has a common base hierarchy of three (3) levels:
+
+`{realm}/v{major_version}/{entity_id}/...`
+
+With:
+* `realm` being a unique id for a domain/realm
+* `v{major_version}` is the major version of keelson used
+* `entity_id` being a unique id representing an entity within the realm
+* `...` are specific key levels depending on the messaging pattern, these are further described below.
+
+> **NOTE:** Without exceptions, keys should adhere to `snake_case` style.
+
+## 2. Publish-Subscribe messaging
+
+### 2.1 Specific key-space design
+
+For pub/sub messaging, the lower levels of the key-space has the following levels:
+
+  `.../data/{subject}/{source_id}`
```

Comment:
All the message specifications are in the folder `messages` why not use the word "messages" instead of "data". I think that doing this change would have 2 advantages: 1) it would make clearer where can one find the "data" specifications; and 2) "the word "messages" has a stronger connection to pub/sub messaging than the word "data". Overall it would make it clearer in my opinion.

Reply:
Related to #31

Reply:
Hmm, l see your point, on the other hand, `messages` has a strong connection also to req/rep messaging. I wanted the levels `data` and `rpc` to describe the traffic on the subsequent levels, `rpc` tells me that it is remote procedure calls going, but, yeah, `data` could be better. 

I am not sure it has to directly reference the `messages` and `interface` folders directly though.

Reply:
`pubsub`!

Reply:
Agreed.